### PR TITLE
[WIP] Flexible transformations for data input and allowing pandas input to generate_counterfactuals

### PR DESCRIPTION
**Pull request by amit-sharma**
This is a work in progress. May break some tests, but the idea is to expose the API and get feedback. The main goal is that user always provide input in the original data scale.

1. `DiCE.Data` and `generate_counterfactuals` always accept data.frame inputs in their original scale.
2. All the transformations are done outside DiCE by the the ML model. E.g., 
```
from sklearn.compose import ColumnTransformer

numeric_transformer = Pipeline(steps=[
    ('imputer', SimpleImputer(strategy='median')),
    #('scaler', StandardScaler())
])

categorical_transformer = Pipeline(steps=[
    ('imputer', SimpleImputer(strategy='constant', fill_value='missing')),
    ('onehot', OneHotEncoder(handle_unknown='ignore'))])

transformations = ColumnTransformer(
    transformers=[
        ('num', numeric_transformer, numerical),
        ('cat', categorical_transformer, categorical)])

clf = Pipeline(steps=[('preprocessor', transformations),
                      ('classifier', LGBMClassifier())])
```

The model is simply `clf` and gets passed to `DiCE.Model`

3. Each explainer method is free to use whatever encoding they prefer. E.g., one-hot for gradient-based dice, and label encoding for kd tree or genetic methods. 

The only question is whether tensorflow and pytorch models have a similar transformation function. I wonder if that is best given as input to `DiCE.Model`. See this, for example, https://www.tensorflow.org/tutorials/structured_data/feature_columns

### CODE REVIEW

File: dice_ml/data_interfaces/public_data_interface.py [hunk 1]
```diff
@@ -86,21 +71,21 @@ def __init__(self, params):
...
 
-        self.label_encoded_data = self.data_df.copy()
-
-        for column in self.categorical_feature_names:
-            self.labelencoder[column] = LabelEncoder()
-            self.label_encoded_data[column] = self.labelencoder[column].fit_transform(self.data_df[column])
-
-        #self.train_df, self.test_df = self.split_data(self.data_df)
-        self.train_df = self.data_df
+        # should move the below snippet to gradient based dice interfaces
```
Comment (raam93) [inline dice_ml/data_interfaces/public_data_interface.py:75]: since line 76-88 are for gradient-based dice only, moved them to dice-interfaces

File: dice_ml/diverse_counterfactuals.py [hunk 2]
```diff
@@ -6,132 +6,38 @@
...
+        self.final_cfs_df_sparse = final_cfs_df_sparse
+
         self.final_cfs_list = None
         self.posthoc_sparsity_param = posthoc_sparsity_param # might be useful for future additions
+        self.test_pred = self.test_instance_df[self.data_interface.outcome_name].iloc[0]
         if desired_class == "opposite":
             self.new_outcome = 1.0 - round(self.test_pred)
         else:
             self.new_outcome = desired_class
         self.encoding = encoding
```
Comment (raam93) [inline dice_ml/diverse_counterfactuals.py:23]: removed convert_to_dataframe() methods as this class now takes in CFs in original raw format (as pd.DataFrame)

File: dice_ml/data_interfaces/public_data_interface.py [hunk 3]
```diff
@@ -261,6 +259,10 @@ def get_data_params_for_gradient_dice(self):
         self.ohe_base_df = self.prepare_df_for_ohe_encoding() # base dataframe for doing one-hot-encoding
         # ohe_encoded_feature_names and ohe_base_df are created (and stored as data class's parameters) when get_data_params_for_gradient_dice() is called from gradient-based DiCE explainers
 
+    def get_data_params_for_gradient_dice(self):
```
Comment (raam93) [inline dice_ml/data_interfaces/public_data_interface.py:262]: similarly, create functions that get parameters required for other dice
